### PR TITLE
add template function getOpsOfType

### PR DESCRIPTION
**Pull request by liqiangxl**
This is the first step to solve #946 : Refactor the existing getXops functions into a single templated function to centralize the operation retrieval logic.

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build

File: csrc/ir/utils.h [hunk 1]
```diff
@@ -505,4 +495,24 @@ bool isTensorSize(const Val* val);
...
+template <typename OpType>
+std::vector<OpType*> getOpsOfType(Fusion* fusion) {
+  std::vector<OpType*> ops;
+  for (auto expr : fusion->exprs()) {
+    if (expr->isA<OpType>()) {
+      ops.push_back(expr->as<OpType>());
+    }
+  }
+  return ops;
+}
```
Comment (zasdfgbnm) [inline csrc/ir/utils.h:510]: Should we do this?
```C++
template <typename ... OpTypes>
auto getOpsOfType(Fusion* fusion) {
  using FirstOpType = std::tuple_element_t<0, std::tuple<Types...>>;
  using ExprType = std:: conditional_t<sizeof...(Types) == 1, FirstOpType, Expr>;
  std::vector<ExprType*> ops;
  for (auto expr : fusion->exprs()) {
    if (expr->isOneOf<OpTypes...>()) {
      ops.push_back(expr->as<ExprType>());
    }
  }
  return ops;
}
```
This way, we can remove `getAllTypesOfReductionOps` as well.

Comment (liqiangxl) [reply csrc/ir/utils.h:510]: I considered variadic template. But it requires the caller knows all types of reduction ops, e.g.
`auto reduction_ops = getOpsOfType<ReductionOp, GroupedReductionOp, WelfordOp>(fusion)`. If this is not a problem, we can remove `getAllTypesOfReductionOps`

**Review by zasdfgbnm — APPROVED** [review]

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/ir/utils.h (hunk 1 above)
Comment (liqiangxl) [reply csrc/ir/utils.h]: Actually I can rewrite the `getAllTypesOfReductionOps` function using the variadic template. Thanks for the suggestion!

**Comment by liqiangxl** [comment]
!build

File: csrc/ir/utils.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/ir/utils.h]: This makes sense. Thanks for the clean up!

**Comment by liqiangxl** [comment]
!build

**Pull request merged**